L69-cpp-core-event-manager.md

```diff
@@ -0,0 +1,282 @@
+  // alarms through a timer slack allows the EventManager to perform 
+  // timer-coalescing, which minimizes thread wake-up and has the potential of 
+  // substantially reducing CPU utilization.
+  virtual void ScheduleAlarmInWithSlack(gpr_timespec deadline, std::function<void(bool)> c, Priority priority, TaskKey* key, gpr_timespec slack) = 0;
+};
+```
+
+Note that there will be no native implementation provided by gRPC for the epoll based interface in the first step. Instead, gRPC will provide a default event manager implementation based on libuv. The libuv EM implementation will be applied to all the gRPC languages based on gRPC core, and the users may choose to override it with their custom implementation of the EM interface. gRPC may release more native event manager implementations down the road map.
+```cpp
+// Represents a file descriptor when using libuv.
+class LibuvDescriptorInterface {
+ public:
+  virtual ~LibuvDescriptorInterface();
+
+  // Returns the underlying TCP or UDP handle. Valid until Close() is called
+  // (i.e. not thread-safe with Close()).
+  virtual uv_handle_t fd() const = 0;
+
+  // Retruns the EventManager object where this descriptor is registered.
+  virtual LibuvEventManagerInterface* event_manager() const = 0;
+
+  // Executes all the registered notification closures, and closes the
+  // underlying TCP or UDP handle.
+  //
+  // Note that we cannot release the underlying TCP or UDP handle, since it is
+  // initialized with the libuv event loop of LibuvEventManager, and it is not
+  // thread-safe to access the handle externally.
+  virtual void Close() = 0;
+  virtual bool IsClosed() const = 0;
+
+  // Registers a closure \a upcall to be run immediately in a worker thread of
+  // the LibuvEventManager, after libuv reads from the socket. \a alloc_cb will
+  // be passed to libuv to allocate the read buffer. Note that this will start
+  // a read asynchronously.
+  //
+  // It is an error to call NotifyAfterRead() again before the previous closure
+  // completes.
+  //
+  // If there is more to read on the socket, re-arms a closure after upcall
+  // finishes.
+  virtuar void NotifyAfterRead(std::function<void(LibuvReadStatus)> upcall,
+                       uv_alloc_cb alloc_cb) = 0;
+
+  // Registers a closure \a upcall to be run immediately in a worker thread of
+  // the LibuvEventManager, after libuv writes to the socket. \a info contains
+  // the write message, and target address in cases of UDP. Note that this will
+  // start a write asynchronously.
+  //
+  // It is an error to call NotifyAfterWrite() again before the previous closure
+  // completes.
+  //
+  // If there is more to write, re-arms a closure after upcall finishes.
+  virtual void NotifyAfterWrite(std::function<void(int)> upcall, LibuvWriteInfo info) = 0;
+
+  // Accessor and mutator of arbitrary user-defined data. NOT thread-safe.
+  virtual void* data() = 0;
```

Comment:
What are these member functions enabling? Could, instead, any user-defined data live in the concrete implementation of the interface?

Reply:
It's for program specific contexts associated with the file descriptors.